storage/innobase/fil/fil0crypt.cc

```diff
@@ -232,40 +233,58 @@ fil_space_crypt_t::fil_space_crypt_t(uint new_type, uint new_min_key_version,
       rotate_state(),
       encryption_rotation(encryption_rotation),
       tablespace_key(NULL) {
+  mutex_create(LATCH_ID_FIL_CRYPT_START_ROTATE_MUTEX, &start_rotate_mutex);
+  mutex_create(LATCH_ID_FIL_CRYPT_DATA_MUTEX, &mutex);
+
   key_id = new_key_id;
   if (my_random_bytes(iv, sizeof(iv)) != MY_AES_OK)  // TODO:Robert: This can
                                                      // return error and because
                                                      // of that it should not be
                                                      // in constructor
     type = 0;  // TODO:Robert: This is temporary to get rid of unused variable
                // problem
-  mutex_create(LATCH_ID_FIL_CRYPT_START_ROTATE_MUTEX, &start_rotate_mutex);
-  mutex_create(LATCH_ID_FIL_CRYPT_DATA_MUTEX, &mutex);
-  // locker = crypt_data_scheme_locker; // TODO:Robert: Co to za locker, nie
-  // mogę znaleść jego definicji nawet w mariadb
-  type = new_type;
+  if (new_uuid != nullptr && strlen(new_uuid) > 0) {
+    memcpy(uuid, new_uuid, ENCRYPTION_SERVER_UUID_LEN);
+    uuid[ENCRYPTION_SERVER_UUID_LEN] = '\0';
+  } else {
+    uuid[0] = '\0';
+  }
 
-  if (new_encryption == FIL_ENCRYPTION_OFF ||
-      (Encryption::is_online_encryption_on() == false &&
-       new_encryption == FIL_ENCRYPTION_DEFAULT)) {
+  if (strlen(new_uuid) == 0) {
+    ut_ad(strlen(server_uuid) == 0);
+    key_found = false;
+    min_key_version = ENCRYPTION_KEY_VERSION_INVALID;
+    // This was read - because when creating new crypt data - it means that
+    // uuid is never empty. type will be overwritten by read function
+  } else if (new_encryption == FIL_ENCRYPTION_OFF) {
     type = CRYPT_SCHEME_UNENCRYPTED;
+    key_found = false;
     min_key_version = ENCRYPTION_KEY_VERSION_NOT_ENCRYPTED;
-    key_found = true;
+  } else if (Encryption::is_online_encryption_on() == false &&
+             new_encryption == FIL_ENCRYPTION_DEFAULT) {
+    type = CRYPT_SCHEME_UNENCRYPTED;
+    min_key_version = ENCRYPTION_KEY_VERSION_NOT_ENCRYPTED;
+    key_found =
+        key_operation == FETCH_OR_GENERATE_KEY
+            ? Encryption::
```

Comment:
This is for all tables that have crypt_data written in page0 and online encryption is disabled. This can be due to ENCRYPTION_KEY_ID being specified, but also for tables that were once encrypted by encryption threads and then were decrypted.

Reply:
Why do we care about crypt_data of the tables that are once encrypted and decrypted later. Can't we treat them as garbage and write the new crypt_data with V3 format.

Also am thinking, since ENCRYPTION_KEY_ID is also KEYRING feature, we can disallow upgrade from previous versions. What do you say? Does it simplify?